Mathlib/CategoryTheory/Adjunction/Triple.lean

```diff
@@ -0,0 +1,55 @@
+/-
+Copyright (c) 2024 Dagur Asgeirsson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Dagur Asgeirsson
+-/
+import Mathlib.CategoryTheory.Adjunction.Unique
+import Mathlib.CategoryTheory.Monad.Adjunction
+/-!
+
+# Adjoint triples
+
+This file concerns adjoint triples `F ⊣ G ⊣ H` of functors `F H : C ⥤ D`, `G : D ⥤ C`.
+
+Currently, the only result is that `F` is fully faithful if and only if `H` is fully faithful.
+-/
+
+namespace CategoryTheory.Adjunction
+
+variable {C D : Type*} [Category C] [Category D]
+variable {F H : C ⥤ D} {G : D ⥤ C}
+variable (adj₁ : F ⊣ G) (adj₂ : G ⊣ H)
+
+lemma isIso_unit_iff_isIso_counit : IsIso adj₁.unit ↔ IsIso adj₂.counit := by
+  let adj : F ⋙ G ⊣ H ⋙ G := adj₁.comp adj₂
+  constructor
+  · intro h
+    let idAdj : 𝟭 C ⊣ H ⋙ G := adj.ofNatIsoLeft (asIso adj₁.unit).symm
+    exact adj₂.isIso_counit_of_iso (idAdj.rightAdjointUniq id)
+  · intro h
+    let adjId : F ⋙ G ⊣ 𝟭 C := adj.ofNatIsoRight (asIso adj₂.counit)
+    exact adj₁.isIso_unit_of_iso (adjId.leftAdjointUniq id)
+
+/--
+Given an adjoint triple `F ⊣ G ⊣ H`, the left adjoint `F` is fully faithful if and only if the
+right adjoint `H` is fully faithful.
+-/
+noncomputable def fullyFaithfulEquiv : F.FullyFaithful ≃ H.FullyFaithful where
+  toFun h :=
+    haveI := h.full
```

Comment:
Does it work to say `haveI := h` instead? I'd hope we have the instance from FullyFaithful to Full. And if so, does it work to omit the `haveI` entirely?

Reply:
That doesn't work, and an instance from `FullyFaithful` to `Full` and to `Faithful` wouldn't work either, because `FullyFaithful` is not a class

Reply:
When `Full` was made a `Prop`, the `FullyFaithful` structure was added, which contains the data of a preimage to every morphism in the image of the functor